libass/ass_library.c

```diff
@@ -95,24 +95,13 @@ void ass_set_style_overrides(ASS_Library *priv, char **list)
         *q = strdup(*p);
 }
 
-static int grow_array(void **array, int nelem, size_t elsize)
-{
-    if (!(nelem & 31)) {
-        void *ptr = realloc(*array, (nelem + 32) * elsize);
-        if (!ptr)
-            return 0;
-        *array = ptr;
-    }
-    return 1;
-}
-
 void ass_add_font(ASS_Library *priv, const char *name, const char *data, int size)
 {
-    int idx = priv->num_fontdata;
+    size_t idx = priv->num_fontdata;
     if (!name || !data || !size)
         return;
-    if (!grow_array((void **) &priv->fontdata, priv->num_fontdata,
-                    sizeof(*priv->fontdata)))
+    if (!(priv->num_fontdata & 31) &&
+            !ASS_REALLOC_ARRAY(priv->fontdata, priv->num_fontdata + 32))
```

Comment:
By the way, we may want to switch this to exponential growth while we’re at it. I suppose this is well-meant because there are usually few memory fonts, but if there should be a situation that needs many memory fonts, this will be quite slow. And when there are few fonts, we won’t be wasting much memory anyway. If we start at 32 and keep doubling the size, our memory use will be the same until we hit 65 fonts, which is already a bigger number than this is likely to be in practice.

Reply:
Can we use exponential growth without adding another variable to keep track of current max size? *(not that adding a variable is a problem)*

Reply:
Ah, hmm. Well, _technically,_ we can follow the spirit of the current code and use bit arithmetic:

```c
if (!(priv->num_fontdata & priv->num_fontdata-1))
    // num_fontdata is 0 or exactly a power of two, so it's time for a realloc
    ASS_REALLOC_ARRAY(priv->fontdata, FFMAX(32, priv->num_fontdata * 2)) ...
```

Reply:
Right, with `2 * …` growth all but the first reallocation will happen on a power of two.